src/aind_metadata_upgrader/procedures_upgrade.py

```diff
@@ -0,0 +1,459 @@
+        """Map legacy InjectionMaterials model to current version"""
+
+        new_materials = []
+        if isinstance(old_injection_materials, list):
+            for (
+                injection_material
+            ) in old_injection_materials:  # this wont work like i want, we changed the naming convention
+                if injection_material.get("titer") is not None:
+                    new_materials.append(self.upgrade_viral_material(injection_material))
+
+                elif injection_material.get("concentration") is not None:
+                    new_materials.append(self.upgrade_nonviral_material(injection_material))
+                else:
+                    logging.error(f"Injection material with no titer or concentration {injection_material} passed in")
+        else:
+            logging.error(f"injection materials not a list: {old_injection_materials}")
+
+        logging.info(f"new_materials: {new_materials}")
+        return new_materials
+
+
+class SubjectProcedureModelsUpgrade(BaseModelUpgrade):
+    """Handle upgrades for SubjectProcedure models."""
+
+    def __init__(self, allow_validation_errors=False):
+        """Handle upgrades for SubjectProcedure models"""
+
+        self.allow_validation_errors = allow_validation_errors
+        self.injection_upgrader = InjectionMaterialsUpgrade(allow_validation_errors)
+
+    def upgrade_craniotomy(self, old_subj_procedure: dict):
+        """Map legacy Craniotomy model to current version"""
+
+        craniotomy_dict = {
+            "protocol_id": old_subj_procedure.get("protocol_id", "unknown"),
+            "craniotomy_type": old_subj_procedure.get("craniotomy_type", None),
+            "craniotomy_hemisphere": get_or_default(old_subj_procedure, Craniotomy, "craniotomy_hemisphere"),
+            "bregma_to_lambda_distance": old_subj_procedure.get("bregma_to_lambda_distance", None),
+            "bregma_to_lambda_unit": get_or_default(old_subj_procedure, Craniotomy, "bregma_to_lambda_unit"),
+            "implant_part_number": old_subj_procedure.get("implant_part_number", None),
+            "dura_removed": get_or_default(old_subj_procedure, Craniotomy, "dura_removed"),
+            "protective_material": get_or_default(old_subj_procedure, Craniotomy, "protective_material"),
+            "recovery_time": old_subj_procedure.get("recovery_time", None),
+            "recovery_time_unit": get_or_default(old_subj_procedure, Craniotomy, "recovery_time_unit"),
+            "craniotomy_size": old_subj_procedure.get("craniotomy_size", None),
+        }
+
+        if not craniotomy_dict['craniotomy_type'] and craniotomy_dict['craniotomy_size']:
+            if '3' in craniotomy_dict['craniotomy_size']:
+                craniotomy_dict['craniotomy_type'] = '3 mm'
+            elif '5' in craniotomy_dict['craniotomy_size']:
+                craniotomy_dict['craniotomy_type'] = '5 mm'
+            
+        drop_unused_fields(craniotomy_dict, 'Craniotomy')
```

Comment:
Referring to my "you should not need `drop_unused_fields` comment. In this case `craniotomy_dict` should only have the fields that the new `Craniotomy` class is requiring by construction.

Reply:
There are some fields that are existing in the upgrading model, but missing in the current model. this check allows those upgrades to still succeed without updating the dict once we pin the schema version to a more modern version.

Reply:
To be more clear, the field `craniotomy_size` is no longer part of the `Craniotomy` model in our most modern version of schema, but still exists in the pinned version of schema for the upgrader. The goal with this line is to future-proof updates to the schema version pin, so that if we ever upgraded it any fields that have been deprecated and removed would be automatically dropped, and not reach construction in the case of an invalid model.